### ISSUE TITLE
Pullback with Functions returning Dictionaries seems to be broken

### ISSUE THREAD

**Issue by ntausend** [FluxML/Zygote.jl#1200]
Hey, I found a weird behavior in having Dictionary types in pullbacks. Consider the two test functions defined below:
```
function test1(di)
    da = di
    da[1] = 5*da[2]
    return da
end

function test2(di)
    da = Dict(1 => di[1], 2 => di[2])
    da[1] = 5*da[2]
    return da
end
```

Both functions having the same result, however the function `test1` modifies the input dictionary in-place while `test2` creates a new dictionary which is then returned.
Using the pullback function at  `d = Dict(1 => el1, 2 => el2)`  one would expect the pullback to be something like (represented as a jacobian)
```
( 0  5 )
( 0  1 )
```
However, executing the code
```
d = Dict(1 => 3, 2 => 5)
_, back_j = Zygote.pullback(testj, d)
back_jx = x -> back_j(x)[1]

display(back_jx(Dict(1 => 3, 2 => 5))
```
for `j` either 1 or 2  gives a different result. In the case of the in-place function (`j=1`) one finds that `back_1x` represents a identity while for `j=2`, `back_2x` is `nothing`. The only way i can explain that behavior is that `Zygote.pullback` tries to differentiate the pointer returned by the test functions, which in the case of `j=1` depends trivially on the input while in the case of `j=2` has no dependence at all on the input dictionary. Instead differentiating the fields would be a more expected behavior.

This picture is supported by the observation, that for `j=2` the same behavior is found using an array as an input instead of an dictionary.

Thanks for any help in advance.

**Comment by ToucheSir** [comment]
Adding a `rrule` for the `Dict` constructor (captured in https://github.com/JuliaDiff/ChainRules.jl/pull/648) makes 2 consistent with 1 (both are now the identity), but still does not induce the same behaviour as one gets with an Array.

**Comment by CarloLucibello** [comment]
```julia
julia> function test3(di)
             da = Dict(1 => 5*di[2], 2 => di[2])
             return da
          end

julia> function test4(di)
             return (a = 5*di[2], b = di[2])
          end

julia> d = Dict(1 => 3, 2 => 5);

julia> gradient(d -> test4(d).a + test4(d).b, d)
(Dict{Any, Any}(2 => 6.0),)

julia> gradient(d -> test3(d)[1] + test3(d)[2], d)
(Dict{Any, Any}(2 => 6.0),)

julia> gradient(d -> test2(d)[1] + test2(d)[2], d)
(Dict{Any, Any}(2 => 6.0),)

julia> gradient(d -> test1(d)[1] + test1(d)[2], d)
(Dict{Any, Any}(2 => 6.0),)
```

**Comment by ToucheSir** [comment]
Not sure, debugging mutating functions with Zygote is unfortunately a pain because you can't set breakpoints and the debugger is liable to blow up if you look at it funny.

If I had to guess, the pullback for `getindex(::Dict, ::Int)::Int` is capturing the original Dict and thus accumulates to the right place, whereas providing a new Dict as a sensitivity to the pullback does not.

**Comment by ntausend** [comment]
Yes it was all about having the pullback of this kind of function. We wanted to implement some fix-point Iteration using a pullback similar to the example. Meanwhile we moved away from using dictionarys and solved that problem.

**Comment by CarloLucibello** [comment]
> Edit: upon re-reading, I don't actually understand what https://github.com/FluxML/Zygote.jl/issues/1200#issuecomment-1328305784 is trying to show. All the examples there work equally well with gradient and pullback. The original issue seems to be with functions which return a Dict instead of a simple scalar, and I can confirm it's still present on 0.6.51.

yes I just wanted to show that gradient computation works fine, while pullback for dict outputs don't, which is weird because the latter should be an intermediate passage of the former, right?

**Comment by ToucheSir** [comment]
Then I think the difference is what I wrote above but later crossed out while trying to understand the example with `gradient`. Namely that using `getindex` and returning a scalar keeps around the right dict, whereas returning the dict itself doesn't accumulate the right dicts (by reference).
